rts/motoko-rts/src/region.rs

```diff
@@ -0,0 +1,854 @@
+        /// None means that the block is available for (re-)allocation.
+        pub fn set(b: BlockId, r: Option<(RegionId, u16)>) {
+            match r {
+                None => write_u16(index(&b), RegionId::into_u16(None)),
+                Some((r, j)) => {
+                    write_u16(index(&b), RegionId::into_u16(Some(r)));
+                    write_u16(index(&b) + 2, j)
+                }
+            }
+        }
+    }
+
+    pub mod region_table {
+        use crate::region::{RegionId, RegionSizeInPages};
+
+        // invariant (for now, pre-GC integration):
+        //  all regions whose IDs are below the total_allocated_regions are valid.
+        use super::{offset, size};
+        use crate::ic0_stable::nicer::{read_u64, write_u64};
+
+        fn index(r: &RegionId) -> u64 {
+            offset::REGION_TABLE + r.0 as u64 * size::REGION_TABLE_ENTRY as u64
+        }
+
+        /// Some(_) gives the size in pages.
+        /// None means that the region is not in use.
+        pub fn get(r: &RegionId) -> Option<RegionSizeInPages> {
+            RegionSizeInPages::from_u64(read_u64(index(r)))
+        }
+
+        pub fn set(r: &RegionId, s: Option<RegionSizeInPages>) {
+            write_u64(index(r), RegionSizeInPages::into_u64(s))
+        }
+    }
+}
+
+unsafe fn alloc_region<M: Memory>(
+    mem: &mut M,
+    id: u16,
+    page_count: u32,
+    vec_pages: Value,
+) -> Value {
+    let r_ptr = mem.alloc_words(size_of::<Region>());
+
+    // NB. cannot use as_region() here as we didn't write the header yet
+    let region = r_ptr.get_ptr() as *mut Region;
+    (*region).header.tag = TAG_REGION;
+    (*region).header.init_forward(r_ptr);
+    (*region).id = id;
+    // The padding must be initialized with zero because it is read by the compiler-generated code.
+    (*region).zero_padding = 0;
+    (*region).page_count = page_count;
+    init_with_barrier(mem, &mut (*region).vec_pages, vec_pages);
+
+    r_ptr
+}
+
+#[ic_mem_fn]
+pub unsafe fn region_id<M: Memory>(_mem: &mut M, r: Value) -> u32 {
+    let r = r.as_region();
+    (*r).id.into()
+}
+
+// Region manager's total memory size in stable memory, in _pages_.
+#[ic_mem_fn]
+pub unsafe fn region_get_mem_size<M: Memory>(_mem: &mut M) -> u64 {
+    let size = {
+        if let Some(s) = crate::region::REGION_SET_MEM_SIZE {
+            return s;
+        };
+        if crate::region::REGION_MEM_SIZE_INIT {
+            meta_data::size::STATIC_MEM_IN_PAGES as u64
+                + crate::region::REGION_TOTAL_ALLOCATED_BLOCKS as u64
+                    * (meta_data::size::PAGES_IN_BLOCK as u64)
+        } else {
+            // Before initialization of anything, give back zero.
+            0
+        }
+    };
+    size
+}
+
+// Region manager's total memory size in stable memory, in _pages_.
+#[ic_mem_fn]
+pub unsafe fn region_set_mem_size<M: Memory>(_mem: &mut M, size: u64) {
+    crate::region::REGION_SET_MEM_SIZE = Some(size);
```

Comment:
This is never called during stable region support, if I am not mistaken.
Maybe a debug assertion could be useful:
`    assert_eq!(crate::region::REGION_MEM_SIZE_INIT, false);`